runtime/lua/vscode-neovim/api.lua

```diff
@@ -1,133 +1,170 @@
-  return M.notify_extension("range-command", command, "v", line1, line2, pos1, pos2, leaveSelection, { ... })
+---Invoke the callback, called by vscode
+---@param id number callback id
+---@param result any resutl
+---@param is_error boolean is error
+function M.invoke_callback(id, result, is_error)
+  vim.schedule(function()
+    local callback = REQUEST_SATE.callbacks[id]
+    REQUEST_SATE.callbacks[id] = nil
+    if callback then
+      if is_error then
+        callback(result, nil)
+      else
+        callback(nil, result)
+      end
+    end
+  end)
 end
 
----call from vscode to sync viewport with neovim
----@param vscode_topline number the top line of vscode visible range
----@param vscode_endline number the end line of vscode visible range
-function M.scroll_viewport(vscode_topline, vscode_endline)
-  local current_height = vim.api.nvim_win_get_height(0)
-  local new_height = vscode_endline - vscode_topline + 1
-  -- resize height
-  if current_height ~= new_height then
-    vim.api.nvim_win_set_height(0, new_height)
+---Run action, this function is asynchronous
+---@param name string The action name, generally a vscode command
+---@param opts? table Optional options table, all fields are optional
+---            - args: (table) Optional arguments for the action
+---            - range: (table) Specific range for the action. In visual mode, this parameter is generally not required.
+---                     The format is consistent with the LSP Protocol.
+---            - line_range: (table) An array [start_line, end_line], which has the same effect as range. Lines are zero indexed
+---            - leave_selection: (boolean) Whether to preserve the selected range, only valid when range or line_range is specified
+---            - callback: (function(err: string|nil, ret: any))
+---                        Optional callback function to handle the action result.
+---                        The first argument is the error message, and the second is the result.
+---                        If no callback is provided, any error message will be shown as a notification in VSCode.
+function M.action(name, opts)
+  vim.validate({
+    name = { name, "string" },
+    opts = { opts, "table", true },
+  })
+  opts = opts or {}
+  vim.validate({
+    ["opts.callback"] = { opts.callback, "f", true },
+    ["opts.args"] = { opts.args, "t", true },
+    ["opts.range"] = { opts.range, "t", true },
+    ["opts.line_range"] = { opts.lien_range, "t", true },
+    ["opts.leave_selection"] = { opts.leave_selection, "b", true },
+  })
+  opts._ = 1
+  if opts.range and opts.line_range then
+    error([[Cannot specify both "range" and "line_range"]])
   end
-
-  local top_line = vim.fn.line("w0")
-  local diff = top_line - vscode_topline
-
-  if diff ~= 0 and (vscode_topline > 0) then
-    vim.fn.winrestview({
-      topline = vscode_topline,
-    })
+  if opts.callback then
+    opts.callback = add_callback(opts.callback)
```

Comment:
Would it be better practice to separately store the id rather than overwriting the function?

Reply:
Functions cannot be passed in RPC